ex8/k_sawada/my_functions/hmm.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+
+
+class HMM:
+    def __init__(self, states, initial_state_probability,
+                 transition_probability, output_probability):
+        """HMM
+
+        Args:
+            states (int):
+                number of states
+            initial_state_probability (ndarray, axis=(state)):
+                initial state probability array
+            transition_probability (ndarray, axis=(before, after)):
+                transition probability matrix
+            output_probability (ndarray, axix=(state, output label)):
+                output probability matrix
+        """
+        self.states = states
+        self.initial_state_probability = initial_state_probability
+        self.transition_probability = transition_probability
+        self.output_probability = output_probability
+    
+    def forward(self, outputs):
+        """likelihoods based on Forward algorithm
+        likelihoods is equal to joint probability
+
+        Args:
+            outputs (ndarray, axis=(time)):
+                outputs series
+
+        Returns:
+            float:
+                likelihoods
+        """
+        length = len(outputs)
+        alpha = np.zeros((length, self.states))
+        
+        # base stage
+        alpha[0] = self.initial_state_probability * \
+            self.output_probability[:, outputs[0]]
+        
+        # recursion stage
+        for t in range(1, length):
+            alpha[t] = (alpha[t - 1].reshape((1, self.states)) @
+                        self.transition_probability).flatten() * \
+                self.output_probability[:, outputs[t]]
```

Comment:
自分でも書いたところですけど、単にこう書いても動く気がしますね
```suggestion
            alpha[t] = (alpha[t - 1] @ self.transition_probability) * self.output_probability[:, outputs[t]]
```

Reply:
確かに…! $(AB)^T = B^TA^T$ ってことですね.